src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -3299,6 +3300,271 @@ tx_gateway_frontend::get_all_transactions() {
+                  "started "
+                  "on the previous term and we can not get current state",
+                  tx.id);
+                co_return tx_errc::invalid_txn_state;
+            }
+        }
+
+        auto r1 = co_await fetch_tx(tx.id, tx.etag);
+        if (!r1) {
+            vlog(
+              txlog.warn,
+              "can't find old tx:{} etag:{} during describe_tx call",
+              tx.id,
+              tx.etag);
+            // Return possibly stale data
+            co_return tx;
+        }
+
+        auto old_tx = r1.value();
+
+        if (old_tx.pid != tx.pid) {
+            vlog(
+              txlog.warn,
+              "fetched pid:{} doesn't match local tx's pid:{}",
+              old_tx.pid,
+              tx.pid);
+            co_return tx_errc::unknown_server_error;
+        }
+
+        if (
+          old_tx.status != tm_transaction::tx_status::ongoing
+          && old_tx.status != tm_transaction::tx_status::ready) {
+            vlog(
+              txlog.warn,
+              "fetched tx:{} pid:{} etag:{} has unexpected status {}",
+              tx.id,
+              old_tx.pid,
+              old_tx.etag,
+              old_tx.status);
+            co_return tx_errc::invalid_txn_state;
+        }
+
+        old_tx.etag = term;
+        auto r2 = co_await stm->update_tx(old_tx, term);
+        if (!r2) {
+            vlog(txlog.info, "got {} on updating fetched tx", r2.error());
+            co_return tx_errc::not_coordinator;
+        }
+        tx = r2.value();
+    } else if (tx.status == tm_transaction::tx_status::killed) {
+        // We can just return current tx, becasue old tx also should be expired
+        co_return tx;
+    } else if (
+      tx.status == tm_transaction::tx_status::prepared
+      || tx.status == tm_transaction::tx_status::aborting) {
+        if (term != tx.etag) {
+            if (!is_fetch_tx_supported()) {
+                vlog(
+                  txlog.warn,
+                  "tm_stm_cache feature is not supported. The tx({}) has "
+                  "started "
+                  "on the previous term and we can not get current state",
+                  tx.id);
+                co_return tx_errc::invalid_txn_state;
+            }
+        }
+
+        auto r1 = co_await fetch_tx(tx.id, tx.etag);
+        if (!r1) {
+            vlog(
+              txlog.warn,
+              "can't find old tx:{} etag:{} during describe_tx call",
+              tx.id,
+              tx.etag);
+            // Return possibly stale data
+            co_return tx;
+        }
+
+        auto old_tx = r1.value();
+
+        if (old_tx.pid != tx.pid) {
+            vlog(
+              txlog.warn,
+              "fetched pid:{} doesn't match local tx's pid:{}",
+              old_tx.pid,
+              tx.pid);
+            co_return tx_errc::unknown_server_error;
+        }
+
+        if (old_tx.status != tm_transaction::tx_status::ongoing) {
```

Comment:
old_tx may have legit `prepared` and `aborting` state, it happens in the following sequence:

 1. old leader commits a tx
 2. the operation fails but the commit passes (mem state has previous ongoing)
 3. the old leader's tm_stm::apply catches up with the committed batch and updates the state
 4. current new leader fetches the state and sees the `prepared` or `aborting` status

when we fetch `prepared` and `aborting` we should check that the `tx_seq` and status match (otherwise - log and return USE) and then update etag.